Bidirectional search for graphs

andre-dietrich opened the pull request:
# Bidirectional Search - Benchmark

Hi,

we added three little algorithms that perform a bidirectional search for paths in a graph (simple, directed, multi-). The benefits are, that the results are generated in order, starting form the shortest paths and then with increased length (in contrast to the original search algorithms). And our approach seems also to be much faster. We therefore provided a small Benchmark below the examples, that show that our algorithm can also be applied for large graphs (15000 nodes) without using much more memory.

We hope, that our little file can be included without replacing the originals ;)
## 0. Code

``` python
    import networkx as nx

    import inspect
    print inspect.getsource(nx.bidirectional.all_simple_paths)
```

```
def all_simple_paths(G, source, target, cutoff=None):
    """Generate all simple paths in graph G from source to target.

    A simple path is a path with no repeated nodes.

    The result paths are returned ordered by there length, starting
    from the shortest paths.

    Parameters
    ----------
    G : NetworkX graph

    source : node
       Starting node for path

    target : node
       Ending node for path

    cutoff : integer, optional
        Depth to stop the search. Only paths of length <= cutoff are returned.

    Returns
    -------
    path_generator: generator
       A generator that produces lists of simple paths, ordered by their
       length starting from the shortes path. If there are no paths between
       the source and target within the given cutoff the generator
       produces no output.

    Examples
    --------
    >>> import networkx as nx
    >>>
    >>> G = nx.complete_graph(4)
    >>> for path in all_simple_paths(G, source=0, target=3):
    ...     print(path)
    ...
    [0, 3]
    [0, 1, 3]
    [0, 2, 3]
    [0, 2, 1, 3]
    [0, 1, 2, 3]
    >>> paths = nx.bidirectional.all_simple_paths(G, source=0, target=3, cutoff=2)
    >>> print(list(paths))
    [[0, 3], [0, 1, 3], [0, 2, 3]]

    Notes
    -----
    This algorithm constructs sequentially two trees, one from the source
    in direction to the target and one from the target in direction to the
    source. The leaves of both growing trees are continuously compared, if
    there are matching leaves in both trees, a new path is identified.

    This algorithm actually reduces the search space by half, which results
    in a faster identification of simple paths than in the original
    "all_simple_paths" algorithm. But it requires more memory, which can
    have a negative effect on the speed in fully connected graphs.


                           source    
                             /\
                            /\ \
                           / /\ \
                          /   /\ \
                         /\  /\  /\
                        /  \/   /  \
                       /    \  /\   \
                      /_____/_/__\___\_________  meeting points
                     /\     \    /   /\
                    /XX\    /  \/   /XX\
                   /XXXX\   \/  \  /XXXX\
                  /XXXXXX\  /    \/XXXXXX\
                 /XXXXXXXX\ \/   /XXXXXXXX\
                /XXXXXXXXXX\ \  /XXXXXXXXXX\
               /XXXXXXXXXXXX\ \/XXXXXXXXXXXX\
              /XXXXXXXXXXXXXX\/XXXXXXXXXXXXXX\
                           target
    Authors
    -------
    André Dietrich and Sebastian Zug
    {dietrich, zug}@ivs.cs.uni-magdeburg.de

    See Also
    --------
    nx.all_simple_paths, nx.all_shortest_paths, nx.shortest_path,
    nx.bidirectional.all_shortest_paths, nx.bidirectional.has_path
    """

    if source not in G:
        raise nx.NetworkXError('source node %s not in graph'%source)
    if target not in G:
        raise nx.NetworkXError('target node %s not in graph'%target)
    if cutoff is None:
        cutoff = len(G)-1
    if G.is_multigraph():
        return _all_simple_paths_multigraph(G, source, target, cutoff=cutoff)
    else:
        return _all_simple_paths_graph(G, source, target, cutoff=cutoff)
```

``` python
    print inspect.getsource(nx.bidirectional._all_simple_paths_graph)
```

```
def _all_simple_paths_graph(G, source, target, cutoff):

    if cutoff < 1:
        return

    tree = [{(source,)}, {(target,)}]

    for i in range(cutoff):      

        tree1  = tree[i%2]
        tree2  = tree[(i+1)%2]
        temp   = set()

        leaves = {x[-1] for x in iter(tree1)} # only to reduce some effort        
        for path in iter(tree2):
            for s in iter(G[path[-1]]):
                if s not in path:
                    if s in leaves:
                        for _path in iter(tree1):
                            if s == _path[-1]:
                                if not set(_path).intersection(path):
                                    if i % 2:
                                        yield list(path)  + [x for x in reversed(_path)]
                                    else:
                                        yield list(_path) + [x for x in reversed(path)]
                    temp.add(path + (s,))

        tree[(i+1)%2] = temp
```
## 1. Examples

``` python
    %pylab inline
    import matplotlib.pyplot as plt
    plt.rcParams['figure.figsize'] = (10.0, 8.0)

    Populating the interactive namespace from numpy and matplotlib

    G = nx.complete_graph(5)  
    nx.draw_circular(G)
```

![output_7_0](https://cloud.githubusercontent.com/assets/3089101/8331677/d917956c-1a89-11e5-816f-40d0d09c2a06.png)
### 1.1 bidirectional.all_simple_paths_tweak

``` python
    for path in nx.bidirectional.all_simple_paths(G, source=0, target=4):
        print(path)
```

```
[0, 4]
[0, 1, 4]
[0, 2, 4]
[0, 3, 4]
[0, 1, 2, 4]
[0, 3, 2, 4]
[0, 2, 1, 4]
[0, 3, 1, 4]
[0, 1, 3, 4]
[0, 2, 3, 4]
[0, 1, 2, 3, 4]
[0, 1, 3, 2, 4]
[0, 3, 1, 2, 4]
[0, 3, 2, 1, 4]
[0, 2, 1, 3, 4]
[0, 2, 3, 1, 4]
```

``` python
    paths = nx.bidirectional.all_simple_paths(G, source=0, target=4, cutoff=2)
    print(list(paths))
```

```
[[0, 4], [0, 1, 4], [0, 2, 4], [0, 3, 4]]
```
### 1.2 bidireactional.all_shortest_paths

``` python
    for path in nx.bidirectional.all_shortest_paths(G, source=0, target=4):
        print(path)
```

```
[0, 4]
```
### 1.3 bidirectional.has_path

``` python
    print nx.bidirectional.has_path(G, source=0, target=4)
```

```
True
```
## 2. Multigraphs

All algorithms can also run on multigraphs, with the same benefits

``` python
    Gm = nx.MultiGraph()
    Gm.add_nodes_from([2,3])
    Gm.add_nodes_from([2,3])
    H=nx.random_geometric_graph(10,0.2, 2)
    H.add_path([0,1,2,3,4,5,6,7,8,9])
    Gm.add_nodes_from(H)
    Gm.add_node(H)

    Gm.add_edges_from(H.edges())
    Gm.add_edges_from(H.edges())

    nx.draw_circular(Gm)
```

![output_16_0](https://cloud.githubusercontent.com/assets/3089101/8331699/fbaf6ce4-1a89-11e5-9ddd-9057595374d6.png)
### 2.1 All simple paths ... produce the same results in order

``` python
    for path in enumerate(nx.bidirectional.all_simple_paths(Gm, 1, 8, cutoff=4)):
        print path
```

```
(0, [1, 3, 8])
(1, [1, 3, 8])
(2, [1, 3, 8])
(3, [1, 3, 8])
(4, [1, 0, 9, 8])
(5, [1, 0, 9, 8])
(6, [1, 0, 9, 8])
(7, [1, 0, 9, 8])
(8, [1, 0, 9, 8])
(9, [1, 0, 9, 8])
(10, [1, 0, 9, 8])
(11, [1, 0, 9, 8])
(12, [1, 2, 3, 8])
(13, [1, 2, 3, 8])
(14, [1, 2, 3, 8])
(15, [1, 2, 3, 8])
(16, [1, 2, 3, 8])
(17, [1, 2, 3, 8])
(18, [1, 2, 3, 8])
(19, [1, 2, 3, 8])
(20, [1, 2, 6, 7, 8])
(21, [1, 2, 6, 7, 8])
(22, [1, 2, 6, 7, 8])
(23, [1, 2, 6, 7, 8])
(24, [1, 2, 6, 7, 8])
(25, [1, 2, 6, 7, 8])
(26, [1, 2, 6, 7, 8])
(27, [1, 2, 6, 7, 8])
(28, [1, 2, 6, 7, 8])
(29, [1, 2, 6, 7, 8])
(30, [1, 2, 6, 7, 8])
(31, [1, 2, 6, 7, 8])
(32, [1, 2, 6, 7, 8])
(33, [1, 2, 6, 7, 8])
(34, [1, 2, 6, 7, 8])
(35, [1, 2, 6, 7, 8])
```

``` python
    for path in enumerate(nx.all_simple_paths(Gm, 1, 8, cutoff=4)):
        print path
```

```
(0, [1, 0, 9, 8])
(1, [1, 0, 9, 8])
(2, [1, 0, 9, 8])
(3, [1, 0, 9, 8])
(4, [1, 0, 9, 8])
(5, [1, 0, 9, 8])
(6, [1, 0, 9, 8])
(7, [1, 0, 9, 8])
(8, [1, 2, 3, 8])
(9, [1, 2, 3, 8])
(10, [1, 2, 3, 8])
(11, [1, 2, 3, 8])
(12, [1, 2, 6, 7, 8])
(13, [1, 2, 6, 7, 8])
(14, [1, 2, 6, 7, 8])
(15, [1, 2, 6, 7, 8])
(16, [1, 2, 6, 7, 8])
(17, [1, 2, 6, 7, 8])
(18, [1, 2, 6, 7, 8])
(19, [1, 2, 6, 7, 8])
(20, [1, 2, 3, 8])
(21, [1, 2, 3, 8])
(22, [1, 2, 3, 8])
(23, [1, 2, 3, 8])
(24, [1, 2, 6, 7, 8])
(25, [1, 2, 6, 7, 8])
(26, [1, 2, 6, 7, 8])
(27, [1, 2, 6, 7, 8])
(28, [1, 2, 6, 7, 8])
(29, [1, 2, 6, 7, 8])
(30, [1, 2, 6, 7, 8])
(31, [1, 2, 6, 7, 8])
(32, [1, 3, 8])
(33, [1, 3, 8])
(34, [1, 3, 8])
(35, [1, 3, 8])
```
### 2.2 but there is a difference in all shortest paths ... our approach seems to find all paths, while the original implimentation does not

``` python
    for path in enumerate(nx.bidirectional.all_shortest_paths(Gm, 1, 8)):
        print path
```

```
(0, [1, 3, 8])
(1, [1, 3, 8])
(2, [1, 3, 8])
(3, [1, 3, 8])
```

``` python
    for path in enumerate(nx.all_shortest_paths(Gm, 1, 8)):
        print path
```

```
(0, [1, 3, 8])
```
## 3.Benchmark

``` python
    import time
    import random

    # trying out a huge graph...
    G1 = nx.dense_gnm_random_graph(15000,15000,2)
```
### 3.1 Speed - has_path

unfortunately, our approach for has_path seems to be slower than the original approach...

``` python
    times1, times2 = [], []
    for i in range(15000):
        source = random.randint(0, len(G1)-1)
        target = random.randint(0, len(G1)-1)

        # tweak
        t = time.time()
        nx.bidirectional.has_path(G1, source, target)
        times2.append(time.time()-t)

        # original
        t = time.time()
        nx.has_path(G1, source, target)
        times1.append(time.time()-t)


    print "original:\t", mean(times1)
    print "bidirectional:\t", mean(times2)
```

```
original:   0.000111669810613
bidirectional:  0.000192139514287
```

``` python
    n, bins, patches = plt.hist(times1, bins=50, normed=1, histtype='bar', rwidth=0.8, label="original")
    n, bins, patches = plt.hist(times2, bins=200, normed=1, histtype='bar', rwidth=0.8, label="bidirectional")
    plt.xlim(0, 0.002)
    plt.ylim(0, 5000)
    plt.legend()
    plt.title("Comparison - has_path")

    <matplotlib.text.Text at 0x7f5f72630ed0>
```

![output_28_1](https://cloud.githubusercontent.com/assets/3089101/8331715/0d70169a-1a8a-11e5-9938-8db167cedce7.png)
### 3.2 Speed - Shortest Paths

... but in the shortest path approach we win ... as you see in the diagrams below, there is even three outliers  below that are below the average of original shortest path search. By the way, both diagrams depict the relation of resulting path lengths and time for their identification... I actually do not know why there are these two chunks in the original approach...

``` python
    times1, times2     = [], []
    results1, results2 = [], []
    lengths1, lengths2 = [], []
    for i in range(5000):
        source = random.randint(0, len(G1)-1)
        target = random.randint(0, len(G1)-1)

        if not nx.has_path(G1, source, target):
            continue

        # bidirectional
        t = time.time()
        res = list(nx.bidirectional.all_shortest_paths(G1, source, target))
        times2.append(time.time()-t)
        results2.append(len(res))
        lengths2.append(len(res[0]))

        # original
        t = time.time()
        res = list(nx.all_shortest_paths(G1, source, target))
        times1.append(time.time()-t)
        results1.append(len(res))
        lengths1.append(len(res[0]))

        #print i, times1[-1], times2[-1]
        #time.sleep(1)


    print "original:\t", mean(times1)
    print "bidirectional:\t", mean(times2)
```

```
original:   0.024720553465
bidirectional:  0.000553524412061
```

``` python
    plt.plot(lengths1, times1, 'bs')
    plt.xlabel("Number of found paths")
    plt.ylabel("Time in seconds")
    plt.title("Original: Time vs. Number of shortest paths found")

    <matplotlib.text.Text at 0x7f5f72531850>
```

![output_32_1](https://cloud.githubusercontent.com/assets/3089101/8331730/1bfb0df0-1a8a-11e5-9253-c712e1ca1519.png)

``` python
    plt.plot(lengths2, times2, 'gs')
    plt.xlabel("Number of found paths")
    plt.ylabel("Time in seconds")
    plt.title("Bidirectional: Time vs. Number of shortest paths found")

    <matplotlib.text.Text at 0x7f5f711de190>
```

![output_33_1](https://cloud.githubusercontent.com/assets/3089101/8331740/26962204-1a8a-11e5-8f38-212a5f11e6a8.png)

``` python
    n, bins, patches = plt.hist(times1, bins=50, normed=1, color="b", histtype='bar', rwidth=0.8, label="original")
    n, bins, patches = plt.hist(times2, bins=50, normed=1, color="g", histtype='bar', rwidth=0.8, label="bidirectional")
    plt.legend()
    plt.title("Comparison - all_shortest_paths")

    <matplotlib.text.Text at 0x7f5f71210610>
```

![output_34_1](https://cloud.githubusercontent.com/assets/3089101/8331744/2efccff6-1a8a-11e5-922b-3c1ed4a426df.png)
### 2.3 Speed - All Simple Paths

Next to the benefit of finding all simple paths in order... the bidirectional search seems also to be much faster... The following code measures the time it takes for generating all simple paths with both approaches. Whereby the cutoff is increased step by step resulting in very huge results sets...

``` python
    times1, times2     = [], []
    results1, results2 = [], []
    print "Cutoff\t#\tTimes original\t\tTimes bidirectional"
    for i in range(31):
        t = time.time()
        res = list(nx.bidirectional.all_simple_paths(G1, 1, 4999, cutoff=i))
        times2.append(time.time()-t)
        results2.append(len(res))    

        # original
        t = time.time()
        res = list(nx.all_simple_paths(G1, 1, 4999, cutoff=i))
        times1.append(time.time()-t)
        results1.append(len(res))

        print i, "\t", results1[-1], "\t",times1[-1], "\t",times2[-1]
        time.sleep(1)
```

```
Cutoff  #   Times original      Times bidirectional
0   0   9.05990600586e-06   2.28881835938e-05
1   0   1.21593475342e-05   4.60147857666e-05
2   0   5.00679016113e-05   0.000115156173706
3   1   3.79085540771e-05   8.9168548584e-05
4   1   4.31537628174e-05   9.41753387451e-05
5   1   0.000125169754028   0.000171184539795
6   1   0.000373125076294   0.00026798248291
7   1   0.000718832015991   0.000355005264282
8   1   0.00141310691833    0.000416040420532
9   1   0.00272083282471    0.000531196594238
10  1   0.00544095039368    0.000647068023682
11  1   0.00988912582397    0.000939846038818
12  2   0.011785030365  0.00146794319153
13  2   0.02867603302   0.00288987159729
14  4   0.0268549919128     0.00106501579285
15  7   0.0633940696716     0.00675392150879
16  14  0.123420000076  0.00654101371765
17  25  0.230616092682  0.0211839675903
18  56  0.479142904282  0.0115299224854
19  102     0.979547977448  0.0393309593201
20  201     1.95987606049   0.0607471466064
21  436     4.03744196892   0.135717868805
22  872     8.60605311394   0.325693130493
23  1814    16.5568730831   0.766483783722
24  3580    34.1799459457   2.57762503624
25  7227    68.5836060047   5.14452195168
26  14670   142.398111105   23.4591898918
27  29613   310.70114994    45.2433450222
28  59372   599.497899055   167.191638947
29  119355  1300.70439291   289.602411032
30  239761  2149.18087816   615.817314863
```

``` python
    plt.title("Search in graph with 15000 nodes: Time vs. Cutoffs")
    plt.ylabel("Time in seconds (logarithmic scale)")
    plt.xlabel("Cutoffs")
    plt.plot(range(len(times1)), times1, color="b", linestyle='-', marker='o', label="nx.all_simple_paths")
    plt.plot(range(len(times2)), times2, color="g", linestyle='-', marker='o', label="bidirectional.all_simple_paths")
    plt.yscale('log')
    plt.legend(loc="lower right")
    plt.grid(True, which="both")
```

![output_37_0](https://cloud.githubusercontent.com/assets/3089101/8331748/386c5b42-1a8a-11e5-99e7-89a2bfd1fa3f.png)
### A smaller graph

Trying the same speed test with a smaller graph (with relatively more edges) seems to produce be similar results... But as you can see in both diagrams, there is something like a break-even point, where it is acutally better to apply the original graph traversing algortithm. Thus, in both logarithmic scales, the original graph traversing shows a linear increase ... in contrast to our approach ... But only for very large result sets or for highly connected graphs.

``` python
    G2 = nx.dense_gnm_random_graph(100,500,1)
    nx.draw_networkx(G2)
```

![output_39_0](https://cloud.githubusercontent.com/assets/3089101/8331755/3f8efc04-1a8a-11e5-8dfd-0eca925381d1.png)

``` python
    times1, times2     = [], []
    results1, results2 = [], []
    print "Cutoff\tPaths\tTime orig.\t\tTime bidirect."
    for i in range(11):
        res = 0
        t = time.time()
        for _ in nx.bidirectional.all_simple_paths(G2, 1, 99, cutoff=i):
            res += 1
        times2.append(time.time()-t)
        results2.append(res)    

        # original
        res = 0
        t = time.time()
        for _ in nx.all_simple_paths(G2, 1, 99, cutoff=i):
            res += 1
        times1.append(time.time()-t)
        results1.append(res)

        print i, "\t", results1[-1], "\t",times1[-1], "\t",times2[-1],
        time.sleep(1)
        print
```

```
Cutoff  Paths   Time orig.      Time bidirect.
0   0   5.96046447754e-06   1.00135803223e-05
1   0   1.71661376953e-05   4.50611114502e-05
2   3   0.000102996826172   0.000167846679688
3   11  0.000810146331787   0.000355958938599
4   99  0.00350999832153    0.00141596794128
5   981     0.0222702026367     0.0144529342651
6   9459    0.35150885582   0.112168073654
7   89808   2.1227080822    0.789459943771
8   840734  19.211204052    8.78042912483
9   7775454     201.188947201   127.474290848
10  70932787    1477.70097995   1536.45552993
```

``` python
    plt.title("Search graph with 500 nodes: Time vs. Cutoffs")
    plt.ylabel("Time in seconds (logarithmic scale)")
    plt.xlabel("Cutoffs")
    plt.plot(range(len(times1)), times1, linestyle='-', marker='o', label="nx.all_simple_paths")
    plt.plot(range(len(times2)), times2, linestyle='-', marker='o', label="bidirectional.all_simple_paths")
    plt.yscale('log')
    plt.legend(loc="lower right")
    plt.grid(True, which="both")
```

![output_41_0](https://cloud.githubusercontent.com/assets/3089101/8331767/4bd37a76-1a8a-11e5-8765-315da8902098.png)
## Memory

Actually I was expecting much more memory consumed by our approach. As you see in the code, we apply the memit, which measures the "process" and the highest amount of memory consumed. The for-loop is chosen to prevent the storage and thus the (probably) wrong increase of elements from the resulting list. The memory measurement should only contain memory allocated by both generators... As it seems, it is more valuable to store the paths, than the iterators over the nodes...

``` python
    from IPython.utils import io
    import memory_profiler

    def load_ipython_extension(ip):
        ip.define_magic('memit', memory_profiler.magic_memit)


    memory1, memory2   = [], []
    print "Level\tMem orig\tMem bidirect"
    for level in range(11):

        res = 0
        with io.capture_output() as mem:
            %memit -r 3 for _ in nx.all_simple_paths(G2, 1, 99, cutoff=level): res+=1
        mem = str(mem.stdout)  
        pos1 = mem.find("increment:") + 11
        pos2 = mem.rfind(" MiB")
        memory1.append(float(mem[pos1:pos2]))

        #time.sleep(5)

        res = 0
        with io.capture_output() as mem:
            %memit -r 3 for _ in nx.bidirectional.all_simple_paths(G2, 1, 99, cutoff=level): res +=1
        mem = str(mem.stdout)  
        pos1 = mem.find("increment:") + 11
        pos2 = mem.rfind(" MiB")
        memory2.append(float(mem[pos1:pos2]))

        #time.sleep(5)

        print level, "\t", memory2[-1], "\t\t", memory1[-1]
```

```
Level   Mem orig    Mem bidirect
0   0.0         0.0
1   0.0         0.0
2   0.0         0.0
3   0.0         0.0
4   0.0         0.0
5   0.0         0.0
6   0.0         0.0
7   0.0         0.0
8   0.27        0.0
9   8.55        0.0
10  26.18       0.0
```

OrkoHunter commented:
Very nice effort @andre-dietrich. However I personally think creating a wiki or some other page was a good idea for all the benchmarks of your code. Loading this junk every time a new comment is created, sometimes feels weird. Anyways, your call.

OrkoHunter commented on networkx/algorithms/bidirectional.py:
```diff
@@ -0,0 +1,390 @@
+    Examples
+    --------
+    >>> import networkx as nx
+    >>>
```
Probably this empty line is not needed.

OrkoHunter commented on networkx/algorithms/bidirectional.py:
```diff
@@ -0,0 +1,390 @@
+#    BSD license.
+import networkx as nx
+__author__ = """\n""".join(['André Dietrich <dietrich@ivs.cs.uni-magdeburg.de>',
+                            'Sebastian Zug <zug@ivs.cs.uni-magdeburg.de>'])
```
What about keeping this variable only in the comment above? `__author__` is not being used now-a-days.

OrkoHunter commented on networkx/algorithms/bidirectional.py:
```diff
@@ -0,0 +1,390 @@
+    
+    This algorithm actually reduces the search space by half, which results
+    in a faster identification of simple paths than in the original
+    "all_simple_paths" algorithm. But it requires more memory, which can
```
Consider making `original "all_simple_paths"` a link to the algorithm.

OrkoHunter commented on networkx/algorithms/bidirectional.py:
```diff
@@ -0,0 +1,390 @@
+                 /XXXXXXXX\ \/   /XXXXXXXX\
+                /XXXXXXXXXX\ \  /XXXXXXXXXX\
+               /XXXXXXXXXXXX\ \/XXXXXXXXXXXX\
+              /XXXXXXXXXXXXXX\/XXXXXXXXXXXXXX\
```
Nice one! :)

andre-dietrich replied:
changed it according to your remarks ...

chebee7i replied:
Neat pic. Let's make sure that Sphinx renders it correctly. We probably will need to mark it as a fixed-width font section.

OrkoHunter replied:
@chebee7i Time to use the system of maintaining fork of `readthedocs.org`.

chebee7i commented:
Looks very nice! I won't have time to review this today, but hopefully we can get it reviewed soon!

MridulS commented:
Please do add tests for this, and try to format your code according to PEP8 :)

andre-dietrich commented:
Hi, everything should now also be formatted according to PEP8, I also included additional tests (whereby I reused already existing ones). The tests were actually a good hint, because I missed directed graphs, which are now also supported. Thus, there are now three additional functions.

OrkoHunter commented on networkx/algorithms/bidirectional.py:
```diff
@@ -0,0 +1,508 @@
+    [0, 1, 3]
+    [0, 2, 3]
+    [0, 2, 1, 3]
+    [0, 1, 2, 3]
```
This is causing a doctest failure on travis.
```
Expected:
    [0, 3]
    [0, 1, 3]
    [0, 2, 3]
    [0, 2, 1, 3]
    [0, 1, 2, 3]
Got:
    [0, 3]
    [0, 1, 3]
    [0, 2, 3]
    [0, 1, 2, 3]
    [0, 2, 1, 3]
```
Consider switching last two lists. Further, after a review note in the PR, instead of making a new commit for every change, you should try other options like `amend` or `squash`.

OrkoHunter commented on networkx/algorithms/tests/test_simple_paths.py:
```diff
@@ -259,3 +259,151 @@ def test_bidirectional_dijkstra_no_path():
+def test_bidirectional_ssp_multigraph():
+    G = nx.MultiGraph()
+    G.add_path([1,2,3])
+    paths = list(nx.bidirectional.all_shortest_paths(G, 1, 4))
```
This will raise a `NetworkXError`, not `NetworkXNotImplemented`.
```
======================================================================
ERROR: test_simple_paths.test_bidirectional_ssp_multigraph
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/nose/tools/nontrivial.py", line 60, in newfunc
    func(*arg, **kw)
  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/networkx/algorithms/tests/test_simple_paths.py", line 409, in test_bidirectional_ssp_multigraph
    paths = list(nx.bidirectional.all_shortest_paths(G, 1, 4))
  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/networkx/algorithms/bidirectional.py", line 284, in all_shortest_paths
    raise nx.NetworkXError('target node %s not in graph' % target)
NetworkXError: target node 4 not in graph
```

andre-dietrich replied:
is running now without throwing an error ...

OrkoHunter commented on networkx/algorithms/tests/test_simple_paths.py:
```diff
@@ -259,3 +259,151 @@ def test_bidirectional_dijkstra_no_path():
+    paths = nx.bidirectional.all_shortest_paths(G, 1, 12)
+    assert_equal(next(paths), [1, 2, 3, 4, 8, 12])
+    assert_equal(next(paths), [1, 5, 9, 10, 11, 12])
+    assert_equal([len(path) for path in nx.bidirectional.all_simple_paths(G, 1, 12)],
```
Longer than 79 characters. It looks like this file needs more such changes. Try running `pep8` over this too.

andre-dietrich replied:
in this case I would leave it as it is, I do not think that readability is increased by splitting the code, since it is a copy the original simple and shortest path tests... see also

215 def test_bidirectional_dijksta_restricted():

which does not meet pep8 too, by the way, there is an (r) missing ;)

OrkoHunter replied:
> I do not think that readability is increased by splitting the code

Yes, I guess you are right.

> see also ... which does not meet pep8 too

Well, this shouldn't encourage you not to follow it.

andre-dietrich replied:
done ;)

OrkoHunter commented on networkx/algorithms/tests/test_simple_paths.py:
```diff
@@ -259,3 +291,157 @@ def test_bidirectional_dijkstra_no_path():
+    assert_equal([len(path) for path in nx.bidirectional.all_simple_paths(G,
+                                                                          1,
+                                                                          12)],
+                 sorted([len(path) for path in nx.all_simple_paths(G, 1, 12)]))
```
I guess this could have been better. You'd love to use `autopep8` in such situations.
```py
assert_equal(
    [len(path) for path in nx.bidirectional.all_simple_paths(G, 1, 12)],
    sorted([len(path) for path in nx.all_simple_paths(G, 1, 12)]))
```

chebee7i commented:
We can iterate on style endlessly :) Has anyone had a chance to review the substance? Unfortunately, it will be a few days before I can even attempt that.

OrkoHunter commented:
@chebee7i For smaller graphs (I checked for around 1000), `nx.bidirectional.all_simple_paths()` takes little bit more time than the original `nx.all_simple_paths()`. The benchmarks claims that this is other way round for very large number of nodes.

andre-dietrich commented:
@OrkoHunter can you post your graph? Is it a fully connected graph? What was the cutoff number you applied?

OrkoHunter commented:
This is pretty much what I tried.
```
In [20]: G = nx.complete_graph(1000)

In [21]: %%timeit
   ....: nx.all_simple_paths(G, 1, 50)
   ....: 
The slowest run took 7.49 times longer than the fastest. This could mean that an intermediate result is being cached 
1000000 loops, best of 3: 1.47 µs per loop

In [22]: %%timeit
   ....: nx.bidirectional.all_simple_paths(G, 1, 50)
   ....: 
The slowest run took 7.39 times longer than the fastest. This could mean that an intermediate result is being cached 
1000000 loops, best of 3: 1.64 µs per loop
```

andre-dietrich commented:
For fully connected graphs we are slower, but I mentioned it in the notes of the function description... The less connected the graph, the better we perform. Otherwise we generate to much overhead...

jtorrents commented:
Very nice indeed! Thanks for working on this @andre-dietrich . I did some quick speed tests with sparse graphs with right skewed degree distributions (a pretty usual use case), and it seems that the proposed implementation is much slower than `nx.all_simple_paths`. See: 

```python
In [135]: %paste
import networkx as nx
from networkx.utils import powerlaw_sequence, create_degree_sequence

def pick_nodes(G):
    u = next(G.nodes_iter())
    v = next(nx.non_neighbors(G, u))
    return u, v

def build_power_law(n):
    deg_seq = create_degree_sequence(n, powerlaw_sequence, 100)
    G = nx.Graph(nx.configuration_model(deg_seq))
    G.remove_edges_from(G.selfloop_edges())
    G = max(nx.biconnected_component_subgraphs(G), key=len)
    G.name = 'Biconnected component of power law configuration model: {0}'.format(n)
    return G
## -- End pasted text --

In [136]: G = build_power_law(40)

In [137]: print(nx.info(G))
Name: Biconnected component of power law configuration model: 40
Type: Graph
Number of nodes: 22
Number of edges: 44
Average degree:   4.0000

In [138]: u, v = pick_nodes(G)

In [139]: timeit paths = list(nx.all_simple_paths(G, u, v))
10 loops, best of 3: 27.4 ms per loop

In [140]: timeit paths = list(nx.shortest_simple_paths(G, u, v))
1 loops, best of 3: 2.66 s per loop

In [141]: timeit paths = list(nx.bidirectional.all_simple_paths(G, u, v))
1 loops, best of 3: 30.6 s per loop
```

@OrkoHunter note that in your speed tests your are not actually computing the paths, you are only creating the generator.

@andre-dietrich  note that we also have a recently added `nx.shortest_simple_paths` which is based on Yen's algorithm for computing k shortest simple paths.

OrkoHunter commented:
@jtorrents Oh yes, I didn't notice. Upon using the generator, the difference is much more clear.

andre-dietrich commented:
@jtorrents Thanks for reviewing ;)

I guess, all simple paths is a bit of problematic in our approach, take a look at the last tow diagrams in our approach. There is some kind of break even point, we are better until a certain cutoff number is reached, after that number we perform worse... That is indicated by the two green curved lines, while the blue line appears to be linear (well in logarithmic scale)... if you test again with some cutoff values, we should perform better, see the code below. 

I took your code with a bit larger graph. The last three lines, show the all_shortest_path approaches. In our solution to this, we could even reduce the effort with a slightly adopted version of our algorithm. Unfortunately, I had to stop your mentioned "nx.shortest_simple_paths", because it took longer than two minutes ...

Cheers,

André

In [2]: G = build_power_law(100)

In [3]: print(nx.info(G))
Name: Biconnected component of power law configuration model: 100
Type: Graph
Number of nodes: 59
Number of edges: 156
Average degree:   5.2881

In [4]: u, v = pick_nodes(G)

In [5]: print len(list(nx.all_simple_paths(G, u, v, cutoff=10)))
278959

In [7]: timeit paths = list(nx.all_simple_paths(G, u, v, cutoff=10))
1 loops, best of 3: 27.2 s per loop

In [8]: timeit paths = list(nx.bidirectional.all_simple_paths(G, u, v, cutoff=10))
1 loops, best of 3: 6.25 s per loop

In [9]: timeit paths = list(nx.all_shortest_paths(G, u, v))
10000 loops, best of 3: 75.5 µs per loop

In [10]: timeit paths = list(nx.bidirectional.all_shortest_paths(G, u, v))
The slowest run took 5.56 times longer than the fastest. This could mean that an intermediate result is being cached 
100000 loops, best of 3: 12.4 µs per loop

In [*]: timeit paths = list(nx.shortest_simple_paths(G, u, v))

andre-dietrich commented:
@jtorrents PS: nx.bidirectional.all_simple_paths(G, u, v, cutoff=10)) returns the results ordered, according to their path length, which from my point of view is a nice feature ;)

jtorrents commented:
@andre-dietrich sorry for being slow on this. The more I look to your implementation the more I like it. However I'm not sure I fully understand why the implementation that we already have for `all_simple_paths` is so much faster without `cutoff` than the proposed here. I think it's because, without cutoff, your implementation is doing `len(G)-1` iterations in the for loop, even if no more paths exist. It would be great to be able to break the for loop when no more paths exists, but I don't see an easy way to do this.

Using a `cutoff` your implementation is significantly faster that we already have, and it's very nice that the paths are yield according to their length. However in the case of `nx.shortest_simple_paths`, it also handles weighted paths, so not all functionality is covered. I think that the best approach is to have single entry points in the public API, and then select the best implementation depending on what the user is asking. So I think that it would be better to substitute or complement the existing functions than to add a new `bidirectional` module for them. We could, for instance, use your implementation of `all_simple_paths` if a `cutoff` is passed as argument; the same for `shortest_simple_paths` if `weight` is `None`, ... 

I'd like to hear more opinions on this (and yours of course). Given than now we are working towards the 2.0 version, we should think about the API and break backwards compatibility if necessary. For instance, I think that `all_simple_paths` should raise `NetworkXNoPath` instead of nothing if no path exists (this is what most path related functions do).

In terms of coding, I think it's better to have both directed and undirected implemented in the same function. Also variable names should be improved. Maybe something like this would do (not tested, sorry):

```python
def _all_simple_paths(G, source, target, cutoff):
    directed = G.is_directed()

    if cutoff < 1:
        return

    tree = [{(source,)}, {(target,)}]

    for i in range(cutoff):
        temp = set()
        tree_forward = tree[i % 2]
        tree_backward = tree[(i + 1) % 2]

        if i % 2:
            if directed:
                neighbors = G.successors_iter
            else:
                neighbors = G.neighbors_iter
        else:
            if directed:
                neighbors = G.predecessors_iter
            else:
                neighbors = G.neighbors_iter

        leaves = {x[-1] for x in tree_forward}  # only to reduce some effort
        for path in tree_backward:
            for s in neighbors(path[-1]):
                if s not in path:
                    if s in leaves:
                        for _path in tree_forward:
                            if s == _path[-1]:
                                if not set(_path).intersection(path):
                                    if i % 2:
                                        yield list(path) + list(reversed(_path))
                                    else:
                                        yield list(_path) + list(reversed(path))
                    temp.add(path + (s,))

        tree[(i + 1) % 2] = temp
```

andre-dietrich commented:
@jtorrents Hi, just changed the code according to your recommendations. 

According to this cutoff problem ... see the last line in the code below ...

``` python
def _all_simple_paths(G, source, target, cutoff):

    if cutoff < 1:
        return

    tree = [{(source,)}, {(target,)}]
    leaves = [{source}, {target}]

    # used if the graph is not directed
    neighbors = lambda node: iter(G[node])

    for i in range(1, cutoff+1):

        if i % 2:
            tree_source = tree[1]
            tree_target = tree[0]
            leaves_ = leaves[0]

            if G.is_directed():
                neighbors = G.successors_iter

        else:
            tree_source = tree[0]
            tree_target = tree[1]
            leaves_ = leaves[1]

            if G.is_directed():
                neighbors = G.predecessors_iter

        temp_tree = set()
        temp_leaves = set()

        for path_s in iter(tree_source):
            for s in neighbors(path_s[-1]):
                if s not in path_s:
                    if s in leaves_:
                        for path_t in iter(tree_target):
                            if s == path_t[-1]:
                                if not set(path_t).intersection(path_s):
                                    found = 1
                                    if i % 2:
                                        yield list(path_t) + \
                                            [x for x in reversed(path_s)]
                                    else:
                                        yield list(path_s) + \
                                            [x for x in reversed(path_t)]
                    temp_tree.add(path_s + (s,))
                    temp_leaves.add(s)

        tree[i % 2] = temp_tree
        leaves[i % 2] = temp_leaves

        print "paths", len(tree[0]), len(tree[1])

```
```
paths 1 2
paths 2 2
paths 2 7
paths 4 7
paths 4 10
paths 8 10
paths 8 22
paths 15 22
paths 15 45
paths 40 45
paths 40 103
paths 77 103
paths 77 213
paths 151 213
paths 151 460
paths 285 460
paths 285 927
paths 611 927
paths 611 1778
paths 1174 1778
paths 1174 3461
paths 2400 3461
paths 2400 7043
paths 4877 7043
```

Thus the "tree" is actually a list of possible sub-paths, which at the next iteration might lead to path from the source to the target... with every iteration the number of sub-paths that have to be stored growths... I guess, that for large cutoffs there is somewhere a point reached, where this overhead reduces the performance ... since the stored paths are currently tuples, this could be reduced with a much nicer data type, another directed graph, a real tree (or a tree composed of dicts with successors and predecessors)... 

There is also this drawback, if multiple paths ... wait a minute ... i had an idea ;)

andre-dietrich commented:
Another "unsuccessfull_leaves" set could decrease the search time, but not tremendously :.(.

``` python
def _all_simple_paths(G, source, target, cutoff):

    if cutoff < 1:
        return

    tree = [{(source,)}, {(target,)}]
    leaves = [{source}, {target}]

    # used if the graph is not directed
    neighbors = lambda node: iter(G[node])

    for i in range(1, cutoff+1):

        if i % 2:
            tree_source = tree[1]
            tree_target = tree[0]
            leaves_ = leaves[0]

            if G.is_directed():
                neighbors = G.successors_iter

        else:
            tree_source = tree[0]
            tree_target = tree[1]
            leaves_ = leaves[1]

            if G.is_directed():
                neighbors = G.predecessors_iter

        temp_tree = set()
        temp_leaves = set()

        unsuccessfull_leaves = set()

        for path_s in iter(tree_source):
            for s in neighbors(path_s[-1]):
                if s not in path_s:
                    if s not in unsuccessfull_leaves:
                        unsuccessfull_leaves.add(s)
                        if s in leaves_:
                            for path_t in iter(tree_target):
                                if s == path_t[-1]:
                                    if not set(path_t).intersection(path_s):
                                        if i % 2:
                                            yield list(path_t) + \
                                                [x for x in reversed(path_s)]
                                        else:
                                            yield list(path_s) + \
                                                [x for x in reversed(path_t)]

                                    try:
                                        unsuccessfull_leaves.remove(s)
                                    except:
                                        pass

                    temp_tree.add(path_s + (s,))
                    temp_leaves.add(s)

        tree[i % 2] = temp_tree
        leaves[i % 2] = temp_leaves
```

jfinkels commented:
An unrelated comment: this is far too many levels of indentation. This should be either refactored into a separate function somehow or possibly adapted to use `for x in (y for y in iterable if ...):` loops instead of the guarded for loops you have now.

andre-dietrich commented:
Hi to everyone

and how about that? I'll put the functions into the simple path module, rename them, do some more documentation and include a link to a newer and more commented version of the benchmark (located at my website)... what would you suggest would be a better for "bidirectional.all_simple_paths" and "bidirectional.all_shortest_paths"?

Regards

jtorrents commented:
@andre-dietrich sorry again for being so slow on this. I'm preparing a review and some benchmarks for this functions. I hope to post them later today or tomorrow.

jtorrents commented:
By the way, can you post the link to the benchmarks on your website?

andre-dietrich commented:
@jtorrents 

http://www.aizac.info/a-bidirectional-graph-search-a-benchmark

andre-dietrich commented:
Hi to everyone,

we improved our code a bit and added an updated benchmark at:

http://www.aizac.info/bi-graph-search-benchmark

as you can see for the all_shortest_paths, we now perform nearly more than 60times faster than the original algorithm...

for all all_simple_paths we perform also better until a certain point but furthermore much better than Yen algorithm (shortest_simple_paths)...

jfinkels commented on networkx/algorithms/bidirectional_path_search.py:
```diff
@@ -0,0 +1,505 @@
+
+        for path_s in iter(tree_source):
+            for s in neighbors(path_s[-1]):
+                if s not in path_s:
```
The level of nesting is still too large to be readable. Perhaps at this point, do

    if s in path_s:
        continue
    if s in leaves_:
        for path_t in tree_target:
            if s != path_t[-1]:
                continue
            if ...

jfinkels commented on networkx/algorithms/bidirectional_path_search.py:
```diff
@@ -0,0 +1,505 @@
+        temp_tree = set()
+        temp_leaves = set()
+
+        for path_s in iter(tree_source):
```
The `iter()` here is unnecessary.

andre-dietrich commented:
#jfinkels  removed the iterators, I thought this way it is a bit faster ...

concerning the continue, I guess every second counts ;) and without continues, the code is slightly more efficient ...

jfinkels commented:
> @jfinkels  removed the iterators, I thought this way it is a bit faster ...

Why would that be faster? The extra function call would make it slower I assume.

> concerning the continue, I guess every second counts ;) and without continues, the code is slightly more efficient ...

Slightly, I guess, in the sense that you avoid an additional `JUMP` instruction (or something like that).

andre-dietrich commented:
@jfinkels yes, due to the additional continue ...

andre-dietrich commented:
Hi to everyone, it is a bit disappointing ... nevertheless, I updated the last versions of our algorithms and added a Jupyter notebook with the benchmarks to:

http://nbviewer.ipython.org/url/gitlab.com/OvGU-ESS/Jupyter-Notebooks/raw/master/bidirectional%20graph%20search%20benchmark.ipynb 

So that everyone can play with it and check, that it is actually better, under certain circumstances of course, than the already included algorithms... I am free for any suggestions, name changes, etc. 

Regards ...

andre-dietrich commented:
Last hi, @jfinkels, @jtorrents, @chebee7i 

before I drop the pull request ;) I changed the some parts of our bidirectional_all_simple_paths, which now performs so good, that I at first thought it was an error ... but even when the other algorithms require 50 seconds to identify all simple paths within a graph, we are still below one second, no matter if it is a large graph with 20.000 or a smaller one with 10 ...

I therefore updated the little benchmark we made, which can be found at:

http://nbviewer.ipython.org/url/gitlab.com/OvGU-ESS/Jupyter-Notebooks/raw/master/bidirectional%20graph%20search%20benchmark.ipynb

In fact all of the developed algorithms perform better than the original ...

OrkoHunter commented:
Hi @andre-dietrich, can you please replace `dict.iteritems()` with `dict.items()` ? That's the only little thing which is causing tests to fail for python3. Rest is all good !

winni2k commented:
I noticed that `all_simple_paths` performs a linear search on the visited path.  I have a PR for a fix at https://github.com/networkx/networkx/pull/3074. I think that may be relevant to this benchmark?
